resources/js/tableofcontent.js

```diff
@@ -1,3 +1,63 @@
+function isHeaderElement(elem) {
+    return elem.nodeName.match(/^H\d+$/) && elem.textContent;
+}
+
+function createHeaderData(elem, i) {
+    const headerText = elem.textContent.trim().replace(/"/g, '\\"');
+    const level = parseInt(elem.nodeName.substr(1));
+    const anchor = `kiwix-toc-${i}`;
+
+    const anchorElem = document.createElement("a");
+    anchorElem.id = anchor;
+
+    /* Mark header content with something we can reference. */
+    elem.insertAdjacentElement("afterbegin", anchorElem);
+
+    return { headerEntry: { text: headerText, anchor, table: [] }, level };
+}
+
+function getDOMHeadersPreorder(elem, array)
+{
+    if (isHeaderElement(elem))
+        array.push(elem);
+
+    Array.from(elem.children).forEach(child => getDOMHeadersPreorder(child, array));
```

Comment:
Why is this better than simple iteration (`for ( const child of elem.children ) { ... }`)? It might matter while you were modifying the DOM while iterating over it but now you can enjoy the separation of DOM traversal and DOM modification.

Reply:
Done